File: src/python/algorithms/graph/wordLadder2/word_ladder2.py
from collections import defaultdict, deque

# BFS (shortest paths) + DFS backtracking => TC: O(N*M*26) SC: O(N*M)
def find_ladders(begin_word, end_word, word_list):
    result = []
    word_set = set(word_list)

    if end_word not in word_set:
        return result

    parents = defaultdict(list)
    visited = {begin_word}
    queue = deque([begin_word])
    found = False

    while queue and not found:
        level_visited = set()

        for _ in range(len(queue)):
            word = queue.popleft()
            arr = list(word)

            for j in range(len(arr)):
                original = arr[j]
                for c in 'abcdefghijklmnopqrstuvwxyz':
                    arr[j] = c
                    next_word = ''.join(arr)

                    if next_word not in word_set:
                        continue

                    if next_word not in visited:
                        queue.append(next_word)
                        level_visited.add(next_word)

                    parents[next_word].append(word)

                    if next_word == end_word:
                        found = True

                arr[j] = original

        visited |= level_visited

    def dfs(word, path):
        if word == begin_word:
            result.append(list(reversed(path)))
            return
        for parent in parents.get(word, []):
            path.append(parent)
            dfs(parent, path)
            path.pop()

    dfs(end_word, [end_word])
    return result


test_cases = [
    ('hit', 'cog', ['hot','dot','dog','lot','log','cog'],
     [['hit','hot','dot','dog','cog'],['hit','hot','lot','log','cog']]),
    ('hit', 'hot', ['hot'],
     [['hit','hot']]),
    ('hit', 'cog', ['hot','dot','dog','lot','log'],
     []),
    ('abc', 'xyz', ['def','ghi','xyz'],
     []),
    ('a', 'c', ['a','b','c'],
     [['a','c']]),
]

for i, (begin, end, words, expected) in enumerate(test_cases, 1):
    result = find_ladders(begin, end, words)
    print(f"Test {i}: Result={result}, Expected={expected}")

File: src/javascript/algorithms/graph/wordLadder2/wordLadder2.js
// BFS (shortest paths) + DFS backtracking => TC: O(N*M*26) SC: O(N*M)
function findLadders(beginWord, endWord, wordList) {
    const result = [];
    const wordSet = new Set(wordList);

    if (!wordSet.has(endWord)) return result;

    const parents = new Map();
    const visited = new Set([beginWord]);
    let queue = [beginWord];
    let found = false;

    while (queue.length && !found) {
        const levelVisited = new Set();
        const nextQueue = [];

        for (const word of queue) {
            const arr = word.split('');
            for (let j = 0; j < arr.length; j++) {
                const original = arr[j];
                for (let c = 97; c <= 122; c++) {
                    arr[j] = String.fromCharCode(c);
                    const nextWord = arr.join('');

                    if (!wordSet.has(nextWord)) continue;

                    if (!visited.has(nextWord)) {
                        nextQueue.push(nextWord);
                        levelVisited.add(nextWord);
                    }

                    if (!parents.has(nextWord)) parents.set(nextWord, []);
                    parents.get(nextWord).push(word);

                    if (nextWord === endWord) found = true;
                }
                arr[j] = original;
            }
        }

        for (const w of levelVisited) visited.add(w);
        queue = nextQueue;
    }

    function dfs(word, path) {
        if (word === beginWord) {
            result.push([...path].reverse());
            return;
        }
        if (!parents.has(word)) return;
        for (const parent of parents.get(word)) {
            path.push(parent);
            dfs(parent, path);
            path.pop();
        }
    }

    dfs(endWord, [endWord]);
    return result;
}

const testCases = [
    {
        beginWord: 'hit', endWord: 'cog',
        wordList: ['hot','dot','dog','lot','log','cog'],
        expected: '[["hit","hot","dot","dog","cog"],["hit","hot","lot","log","cog"]]'
    },
    {
        beginWord: 'hit', endWord: 'hot',
        wordList: ['hot'],
        expected: '[["hit","hot"]]'
    },
    {
        beginWord: 'hit', endWord: 'cog',
        wordList: ['hot','dot','dog','lot','log'],
        expected: '[]'
    },
    {
        beginWord: 'abc', endWord: 'xyz',
        wordList: ['def','ghi','xyz'],
        expected: '[]'
    },
    {
        beginWord: 'a', endWord: 'c',
        wordList: ['a','b','c'],
        expected: '[["a","c"]]'
    },
];

testCases.forEach(({ beginWord, endWord, wordList, expected }, i) => {
    const result = findLadders(beginWord, endWord, wordList);
    console.log(`Test ${i + 1}: Result=${JSON.stringify(result)}, Expected=${expected}`);
});
